File: apps/server/src/newsletter/email.ts
export interface NewsletterEmail {
  to: string;
  subject: string;
  html: string;
  text: string;
  headers?: Record<string, string>;
  tags?: { name: string; value: string }[];
}

export interface EmailDelivery {
  id: string;
}

export type EmailSender = (email: NewsletterEmail, idempotencyKey: string) => Promise<EmailDelivery>;

export class NewsletterConfigurationError extends Error {}

export function createResendSender(environment: NodeJS.ProcessEnv = process.env): EmailSender {
  return async (email, idempotencyKey) => {
    const apiKey = environment.RESEND_API_KEY?.trim();
    const from = environment.NEWSLETTER_FROM?.trim();
    if (!apiKey || !from) {
      throw new NewsletterConfigurationError("Newsletter delivery is not configured");
    }

    const requestBody = JSON.stringify({
      from,
      to: [email.to],
      subject: email.subject,
      html: email.html,
      text: email.text,
      ...(environment.NEWSLETTER_REPLY_TO?.trim()
        ? { reply_to: environment.NEWSLETTER_REPLY_TO.trim() }
        : {}),
      ...(email.headers ? { headers: email.headers } : {}),
      ...(email.tags ? { tags: email.tags } : {}),
    });

    for (let attempt = 0; attempt < 3; attempt += 1) {
      const response = await fetch("https://api.resend.com/emails", {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
          "Idempotency-Key": idempotencyKey,
        },
        body: requestBody,
      });

      const body = (await response.json().catch(() => ({}))) as { id?: string; message?: string; error?: string };
      if (response.ok && body.id) return { id: body.id };

      const retryable = response.status === 429 || response.status >= 500;
      if (!retryable || attempt === 2) {
        throw new Error(body.message || body.error || `Email provider returned ${response.status}`);
      }
      const retryAfterSeconds = Number(response.headers.get("retry-after"));
      const delay = Number.isFinite(retryAfterSeconds) && retryAfterSeconds > 0
        ? retryAfterSeconds * 1000
        : 650 * (attempt + 1);
      await new Promise((resolve) => setTimeout(resolve, delay));
    }

    throw new Error("Email provider retry limit reached");
  };
}

export function escapeHtml(value: string): string {
  return value
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&#039;");
}

function emailLayout(content: string, footer: string): string {
  return `<!doctype html>
<html>
  <body style="margin:0;background:#0b0b10;color:#f5f5f7;font-family:Arial,sans-serif">
    <div style="max-width:620px;margin:0 auto;padding:36px 20px">
      <div style="font-size:12px;font-weight:800;letter-spacing:.16em;text-transform:uppercase;color:#9b87f5;margin-bottom:24px">AI &amp; Tech News</div>
      <div style="background:#15151d;border:1px solid #292936;border-radius:8px;padding:28px">${content}</div>
      <div style="color:#8f8f9c;font-size:12px;line-height:1.6;padding:20px 4px">${footer}</div>
    </div>
  </body>
</html>`;
}

function button(label: string, url: string): string {
  return `<a href="${escapeHtml(url)}" style="display:inline-block;background:#7c5cff;color:#fff;text-decoration:none;font-weight:700;border-radius:5px;padding:13px 20px">${escapeHtml(label)}</a>`;
}

export function confirmationEmail(to: string, confirmUrl: string): NewsletterEmail {
  const content = `
    <h1 style="margin:0 0 14px;font-size:26px;line-height:1.2">Confirm your subscription</h1>
    <p style="color:#c8c8d0;line-height:1.65;margin:0 0 22px">Confirm that you want the most useful AI and technology stories delivered to your inbox.</p>
    ${button("Confirm subscription", confirmUrl)}
    <p style="color:#8f8f9c;font-size:12px;line-height:1.5;margin:22px 0 0">This link expires in 48 hours. If you did not request this email, you can ignore it.</p>`;
  return {
    to,
    subject: "Confirm your AI & Tech News subscription",
    html: emailLayout(content, "You received this confirmation because someone entered this address at aiandtech.news."),
    text: `Confirm your AI & Tech News subscription:\n${confirmUrl}\n\nThis link expires in 48 hours. If you did not request this email, ignore it.`,
    tags: [{ name: "email_type", value: "confirmation" }],
  };
}

export function welcomeEmail(to: string, siteUrl: string, unsubscribeUrl: string): NewsletterEmail {
  const content = `
    <h1 style="margin:0 0 14px;font-size:26px;line-height:1.2">Welcome to AI &amp; Tech News</h1>
    <p style="color:#c8c8d0;line-height:1.65;margin:0 0 22px">Your subscription is confirmed. We will send you a concise daily digest of the AI and technology stories worth knowing.</p>
    ${button("Read the latest stories", siteUrl)}`;
  return {
    to,
    subject: "Welcome to AI & Tech News",
    html: emailLayout(content, `You can <a href="${escapeHtml(unsubscribeUrl)}" style="color:#b7a7ff">unsubscribe at any time</a>.`),
    text: `Welcome to AI & Tech News. Your subscription is confirmed.\n\nRead the latest stories: ${siteUrl}\n\nUnsubscribe: ${unsubscribeUrl}`,
    headers: {
      "List-Unsubscribe": `<${unsubscribeUrl}>`,
      "List-Unsubscribe-Post": "List-Unsubscribe=One-Click",
    },
    tags: [{ name: "email_type", value: "welcome" }],
  };
}

export interface DigestArticle {
  title: string;
  slug: string;
  excerpt: string;
  category: string;
  readingMinutes: number;
}

export function digestSubject(articles: DigestArticle[]): string {
  const lead = articles[0];
  if (!lead) return "Today in AI and technology";
  const rest = articles.length - 1;
  return rest > 0 ? `${lead.title} (+${rest} more)` : lead.title;
}

// Groups articles under their category so the digest reads as sections
// rather than one flat list, preserving the order they were selected in.
export function groupByCategory(articles: DigestArticle[]): { category: string; articles: DigestArticle[] }[] {
  const sections: { category: string; articles: DigestArticle[] }[] = [];
  for (const article of articles) {
    const existing = sections.find((section) => section.category === article.category);
    if (existing) existing.articles.push(article);
    else sections.push({ category: article.category, articles: [article] });
  }
  return sections;
}

export function digestEmail(
  to: string,
  articles: DigestArticle[],
  siteUrl: string,
  unsubscribeUrl: string,
): NewsletterEmail {
  const sections = groupByCategory(articles);

  const sectionsHtml = sections
    .map((section) => {
      const items = section.articles
        .map((article) => {
          const url = `${siteUrl}/article/${encodeURIComponent(article.slug)}`;
          return `<div style="padding:0 0 18px">
        <a href="${escapeHtml(url)}" style="color:#f5f5f7;text-decoration:none;font-size:19px;line-height:1.3;font-weight:800">${escapeHtml(article.title)}</a>
        <span style="color:#8f8f9c;font-size:13px;white-space:nowrap"> (${article.readingMinutes} minute read)</span>
        <p style="color:#b8b8c2;line-height:1.55;margin:7px 0 0">${escapeHtml(article.excerpt)}</p>
      </div>`;
        })
        .join("");
      return `<div style="padding:18px 0 0;border-top:1px solid #292936">
        <div style="color:#9b87f5;font-size:11px;font-weight:800;letter-spacing:.12em;text-transform:uppercase;margin-bottom:12px">${escapeHtml(section.category)}</div>
        ${items}
      </div>`;
    })
    .join("");

  const textArticles = sections
    .map((section) => {
      const items = section.articles
        .map(
          (article) =>
            `${article.title} (${article.readingMinutes} minute read)\n${article.excerpt}\n${siteUrl}/article/${encodeURIComponent(article.slug)}`,
        )
        .join("\n\n");
      return `${section.category.toUpperCase()}\n\n${items}`;
    })
    .join("\n\n");

  const totalMinutes = articles.reduce((sum, article) => sum + article.readingMinutes, 0);
  const content = `
    <h1 style="margin:0 0 8px;font-size:26px;line-height:1.2">Today in AI and technology</h1>
    <p style="color:#c8c8d0;line-height:1.65;margin:0 0 18px">${articles.length} ${articles.length === 1 ? "story" : "stories"} worth knowing, about ${totalMinutes} ${totalMinutes === 1 ? "minute" : "minutes"} of reading.</p>
    ${sectionsHtml}
    <div style="padding-top:14px;border-top:1px solid #292936">${button("See all stories", siteUrl)}</div>`;
  return {
    to,
    subject: digestSubject(articles),
    html: emailLayout(content, `You subscribed at aiandtech.news. <a href="${escapeHtml(unsubscribeUrl)}" style="color:#b7a7ff">Unsubscribe</a>.`),
    text: `Today in AI and technology\n\n${textArticles}\n\nSee all stories: ${siteUrl}\n\nUnsubscribe: ${unsubscribeUrl}`,
    headers: {
      "List-Unsubscribe": `<${unsubscribeUrl}>`,
      "List-Unsubscribe-Post": "List-Unsubscribe=One-Click",
    },
    tags: [{ name: "email_type", value: "daily_digest" }],
  };
}

File: apps/server-go/internal/newsletter/email.go
package newsletter

import (
	"strconv"
	"strings"
)

// Email is Node's NewsletterEmail. Headers keep Node's object key order,
// which is the order they appear in the Resend request body.
type Email struct {
	To      string
	Subject string
	HTML    string
	Text    string
	Headers []Header
	Tags    []Tag
}

type Header struct {
	Name  string
	Value string
}

type Tag struct {
	Name  string
	Value string
}

// DigestArticle is one story in a digest and in the stored edition JSON
// (field order and names are Node's).
type DigestArticle struct {
	Title          string `json:"title"`
	Slug           string `json:"slug"`
	Excerpt        string `json:"excerpt"`
	Category       string `json:"category"`
	ReadingMinutes int    `json:"readingMinutes"`
}

// The templates below are apps/server/src/newsletter/email.ts verbatim,
// including every newline and indentation run inside the template literals.

func emailLayout(content, footer string) string {
	return `<!doctype html>
<html>
  <body style="margin:0;background:#0b0b10;color:#f5f5f7;font-family:Arial,sans-serif">
    <div style="max-width:620px;margin:0 auto;padding:36px 20px">
      <div style="font-size:12px;font-weight:800;letter-spacing:.16em;text-transform:uppercase;color:#9b87f5;margin-bottom:24px">AI &amp; Tech News</div>
      <div style="background:#15151d;border:1px solid #292936;border-radius:8px;padding:28px">` + content + `</div>
      <div style="color:#8f8f9c;font-size:12px;line-height:1.6;padding:20px 4px">` + footer + `</div>
    </div>
  </body>
</html>`
}

func button(label, url string) string {
	return `<a href="` + escapeHTML(url) + `" style="display:inline-block;background:#7c5cff;color:#fff;text-decoration:none;font-weight:700;border-radius:5px;padding:13px 20px">` + escapeHTML(label) + `</a>`
}

func unsubscribeHeaders(unsubscribeURL string) []Header {
	return []Header{
		{Name: "List-Unsubscribe", Value: "<" + unsubscribeURL + ">"},
		{Name: "List-Unsubscribe-Post", Value: "List-Unsubscribe=One-Click"},
	}
}

// WelcomeEmail is Node's welcomeEmail, sent when a pending subscriber confirms.
func WelcomeEmail(to, siteURL, unsubscribeURL string) Email {
	content := `
    <h1 style="margin:0 0 14px;font-size:26px;line-height:1.2">Welcome to AI &amp; Tech News</h1>
    <p style="color:#c8c8d0;line-height:1.65;margin:0 0 22px">Your subscription is confirmed. We will send you a concise daily digest of the AI and technology stories worth knowing.</p>
    ` + button("Read the latest stories", siteURL)
	return Email{
		To:      to,
		Subject: "Welcome to AI & Tech News",
		HTML:    emailLayout(content, `You can <a href="`+escapeHTML(unsubscribeURL)+`" style="color:#b7a7ff">unsubscribe at any time</a>.`),
		Text:    "Welcome to AI & Tech News. Your subscription is confirmed.\n\nRead the latest stories: " + siteURL + "\n\nUnsubscribe: " + unsubscribeURL,
		Headers: unsubscribeHeaders(unsubscribeURL),
		Tags:    []Tag{{Name: "email_type", Value: "welcome"}},
	}
}

// DigestSubject is Node's digestSubject: the lead story's title, with the
// number of further stories.
func DigestSubject(articles []DigestArticle) string {
	if len(articles) == 0 {
		return "Today in AI and technology"
	}
	if rest := len(articles) - 1; rest > 0 {
		return articles[0].Title + " (+" + strconv.Itoa(rest) + " more)"
	}
	return articles[0].Title
}

type digestSection struct {
	category string
	articles []DigestArticle
}

// groupByCategory is Node's groupByCategory: sections in order of first
// appearance, articles in selection order, categories compared exactly.
func groupByCategory(articles []DigestArticle) []digestSection {
	var sections []digestSection
	for _, article := range articles {
		found := false
		for i := range sections {
			if sections[i].category == article.Category {
				sections[i].articles = append(sections[i].articles, article)
				found = true
				break
			}
		}
		if !found {
			sections = append(sections, digestSection{category: article.Category, articles: []DigestArticle{article}})
		}
	}
	return sections
}

func plural(count int, one, many string) string {
	if count == 1 {
		return one
	}
	return many
}

// DigestEmail is Node's digestEmail.
func DigestEmail(to string, articles []DigestArticle, siteURL, unsubscribeURL string) Email {
	sections := groupByCategory(articles)

	var sectionsHTML strings.Builder
	textSections := make([]string, 0, len(sections))
	for _, section := range sections {
		var items strings.Builder
		textItems := make([]string, 0, len(section.articles))
		for _, article := range section.articles {
			url := siteURL + "/article/" + encodeURIComponent(article.Slug)
			minutes := strconv.Itoa(article.ReadingMinutes)
			items.WriteString(`<div style="padding:0 0 18px">
        <a href="` + escapeHTML(url) + `" style="color:#f5f5f7;text-decoration:none;font-size:19px;line-height:1.3;font-weight:800">` + escapeHTML(article.Title) + `</a>
        <span style="color:#8f8f9c;font-size:13px;white-space:nowrap"> (` + minutes + ` minute read)</span>
        <p style="color:#b8b8c2;line-height:1.55;margin:7px 0 0">` + escapeHTML(article.Excerpt) + `</p>
      </div>`)
			textItems = append(textItems, article.Title+" ("+minutes+" minute read)\n"+article.Excerpt+"\n"+url)
		}
		sectionsHTML.WriteString(`<div style="padding:18px 0 0;border-top:1px solid #292936">
        <div style="color:#9b87f5;font-size:11px;font-weight:800;letter-spacing:.12em;text-transform:uppercase;margin-bottom:12px">` + escapeHTML(section.category) + `</div>
        ` + items.String() + `
      </div>`)
		textSections = append(textSections, jsToUpper(section.category)+"\n\n"+strings.Join(textItems, "\n\n"))
	}

	totalMinutes := 0
	for _, article := range articles {
		totalMinutes += article.ReadingMinutes
	}
	content := `
    <h1 style="margin:0 0 8px;font-size:26px;line-height:1.2">Today in AI and technology</h1>
    <p style="color:#c8c8d0;line-height:1.65;margin:0 0 18px">` + strconv.Itoa(len(articles)) + ` ` + plural(len(articles), "story", "stories") +
		` worth knowing, about ` + strconv.Itoa(totalMinutes) + ` ` + plural(totalMinutes, "minute", "minutes") + ` of reading.</p>
    ` + sectionsHTML.String() + `
    <div style="padding-top:14px;border-top:1px solid #292936">` + button("See all stories", siteURL) + `</div>`
	return Email{
		To:      to,
		Subject: DigestSubject(articles),
		HTML:    emailLayout(content, `You subscribed at aiandtech.news. <a href="`+escapeHTML(unsubscribeURL)+`" style="color:#b7a7ff">Unsubscribe</a>.`),
		Text:    "Today in AI and technology\n\n" + strings.Join(textSections, "\n\n") + "\n\nSee all stories: " + siteURL + "\n\nUnsubscribe: " + unsubscribeURL,
		Headers: unsubscribeHeaders(unsubscribeURL),
		Tags:    []Tag{{Name: "email_type", Value: "daily_digest"}},
	}
}
